File: chronicle-pc/pipeline/chronicle_pipeline/path_map.py
"""PARA knowledge path map (v1.10 — dual-read cutover complete).

Knowledge candidates are PARA areas only. Legacy ``kb/notes/`` is retired;
``is_legacy_kb_path`` remains for cutover/410 detection. Journal paths
(``entries/``, ``img/``, ``audio/``, ``notes/``) are unchanged.
"""

from __future__ import annotations

import re
from collections.abc import Iterator
from pathlib import Path
from typing import Any

# Preferred knowledge layout (PARA). Numbers keep sort order in file browsers.
PARA_AREAS: tuple[str, ...] = (
    "00-Inbox",
    "10-Work",
    "20-Personal",
    "30-Knowledge",
    "90-Archive",
)

LEGACY_KB_NOTES = "kb/notes"

# UI "Notes" sections (v1.10): Knowledge Base = 30-Knowledge only;
# Notes = all other PARA areas. Journal is handled separately.
KB_AREA = "30-Knowledge"
NOTES_AREAS: tuple[str, ...] = tuple(a for a in PARA_AREAS if a != KB_AREA)
SECTION_KB = "kb"
SECTION_NOTES = "notes"
VALID_SECTIONS: frozenset[str] = frozenset({SECTION_KB, SECTION_NOTES})

# Hide from app knowledge trees (keep MOCs). Matches Android KnowledgePathMap.
CHROME_BASENAMES: frozenset[str] = frozenset({"CLAUDE.md", ".gitkeep", "README.md"})

# Machine / regenerable dirs — Obsidian should exclude these when opening vault root.
MACHINE_EXCLUDE_DIRS: tuple[str, ...] = (
    "index",
    "brain",
    "_capture",
    "_attachments",
    "entries",
    "img",
    "audio",
    ".stfolder",
    "_staging",
)

_SAFE_REL = re.compile(r"^[A-Za-z0-9._\- /]+$")
_FORBIDDEN = ("..", "\0")


def is_para_prefix(rel: str) -> bool:
    """True if vault-relative path is under a PARA area (or is the area itself)."""
    p = _norm(rel)
    for area in PARA_AREAS:
        if p == area or p.startswith(area + "/"):
            return True
    return False


def is_legacy_kb_path(rel: str) -> bool:
    """True for retired kb/notes paths (cutover/410 detection only)."""
    p = _norm(rel)
    return p == LEGACY_KB_NOTES or p.startswith(LEGACY_KB_NOTES + "/")


def is_knowledge_path(rel: str) -> bool:
    """True if path is under a PARA knowledge area (legacy kb/notes excluded)."""
    return is_para_prefix(rel)


def section_for(rel: str) -> str | None:
    """UI section for a knowledge-relative path: 'kb', 'notes', or None if not knowledge."""
    p = _norm(rel)
    if p == KB_AREA or p.startswith(KB_AREA + "/"):
        return SECTION_KB
    for area in NOTES_AREAS:
        if p == area or p.startswith(area + "/"):
            return SECTION_NOTES
    return None


def default_create_area(section: str) -> str:
    """Default PARA area for a new note in the given UI section."""
    return KB_AREA if section == SECTION_KB else "00-Inbox"


def validate_section(section: str | None) -> str | None:
    """Return normalized section or None. Raise ValueError on garbage values."""
    if section is None or section == "":
        return None
    if section not in VALID_SECTIONS:
        raise ValueError(f"section must be 'kb' or 'notes', got {section!r}")
    return section


def is_chrome_basename(name: str) -> bool:
    return name in CHROME_BASENAMES


def is_chrome_path(rel: str) -> bool:
    return Path(_norm(rel)).name in CHROME_BASENAMES


def path_allowed_for_section(rel: str, section: str | None) -> bool:
    """Hard create-scope: when section is set, path must belong to that section."""
    if section is None:
        return True
    try:
        section = validate_section(section)
    except ValueError:
        return False
    return section_for(rel) == section


def assert_path_allowed_for_section(rel: str, section: str | None) -> None:
    """Raise ValueError if [rel] is outside [section]'s areas."""
    section = validate_section(section)
    if section is None:
        return
    if section_for(rel) != section:
        raise ValueError(
            f"path {rel!r} is outside section {section!r} "
            f"(kb → {KB_AREA}/; notes → Inbox/Work/Personal/Archive)"
        )


def para_area_roots(root: Path) -> list[Path]:
    return [root / area for area in PARA_AREAS]


def knowledge_roots(root: Path, *, existing_only: bool = True) -> list[Path]:
    """Roots to scan for knowledge markdown (PARA only)."""
    roots: list[Path] = []
    for area in PARA_AREAS:
        p = root / area
        if not existing_only or p.is_dir():
            roots.append(p)
    return roots


def _norm(path: str) -> str:
    p = path.strip().lstrip("/").replace("\\", "/")
    while "//" in p:
        p = p.replace("//", "/")
    return p


def normalize_api_path(path: str) -> str:
    """
    Normalize a serve/API path to a vault-relative knowledge path.

    Accepts:
      - PARA: 10-Work/ResumePoints/foo.md
      - legacy full: kb/notes/foo.md (returned as-is for 410 detection)
      - bare relative: ResumePoints/foo.md → 10-Work/…; else → 00-Inbox/…
    """
    p = _norm(path)
    if p.startswith("kb/notes/"):
        return p
    if p == "kb/notes":
        raise ValueError("path must be a .md file")
    if is_para_prefix(p):
        return p
    # Bare relative → PARA (dual-read alias retired)
    if p.startswith("ResumePoints/"):
        return f"10-Work/{p}"
    return f"00-Inbox/{p}"


def validate_knowledge_rel(rel: str) -> str:
    """Validate and return normalized vault-relative PARA knowledge .md path.

    Raises ValueError for invalid paths. Legacy ``kb/notes/…`` raises with a
    cutover hint (callers that need HTTP 410 should check ``is_legacy_kb_path``
    before validating).
    """
    p = normalize_api_path(rel)
    if is_legacy_kb_path(p):
        raise ValueError(
            "legacy kb/notes/ path retired; run chronicle cutover-kb "
            "and use a PARA path"
        )
    if any(f in p for f in _FORBIDDEN) or p.startswith("../") or "/../" in p:
        raise ValueError("invalid note path")
    if not _SAFE_REL.match(p):
        raise ValueError("note path has invalid characters")
    if not p.endswith(".md"):
        raise ValueError("note path must end with .md")
    if not is_knowledge_path(p):
        raise ValueError("path is not under a PARA knowledge area")
    return p


def abs_under_root(root: Path, rel: str) -> Path:
    """Resolve vault-relative path; raise ValueError if it escapes root."""
    base = root.resolve()
    target = (root / rel).resolve()
    if not target.is_relative_to(base):
        raise ValueError("note path escapes vault root")
    return target


def candidate_read_paths(root: Path, rel: str) -> list[Path]:
    """
    Ordered candidates for reading a note (PARA only).

    Exact vault-relative path first. ResumePoints: bare ``ResumePoints/X.md``
    probes ``10-Work/ResumePoints/X.md`` via normalize. No legacy kb/notes peers.
    """
    rel = normalize_api_path(rel)
    seen: set[Path] = set()
    out: list[Path] = []

    def add(p: Path) -> None:
        try:
            rp = p.resolve()
        except OSError:
            return
        if rp in seen:
            return
        seen.add(rp)
        out.append(p)

    if is_legacy_kb_path(rel):
        # Retired — no candidates (API returns 410)
        return out

    add(root / rel)
    return out


def same_suffix_peers(root: Path, rel: str) -> list[str]:
    """Existing PARA path for [rel] (self only after dual-read cutover)."""
    rel = normalize_api_path(rel)
    peers: list[str] = []
    if is_legacy_kb_path(rel):
        return peers
    if is_para_prefix(rel) and (root / rel).is_file():
        peers.append(rel)
    return peers


def find_dual_copy_pairs(root: Path) -> list[dict[str, str]]:
    """Report leftover PARA + legacy same-suffix copies (doctor / pre-cutover)."""
    pairs: list[dict[str, str]] = []
    legacy_root = root / "kb" / "notes"
    if not legacy_root.is_dir():
        return pairs
    for path in sorted(legacy_root.rglob("*.md")):
        if path.name.startswith(".") or ".sync-conflict" in path.name:
            continue
        if is_chrome_basename(path.name):
            continue
        rel = vault_rel(root, path)
        suffix = rel[len(LEGACY_KB_NOTES) + 1 :]
        for area in PARA_AREAS:
            para = root / area / suffix
            if para.is_file():
                pairs.append(
                    {
                        "suffix": suffix,
                        "para": f"{area}/{suffix}",
                        "legacy": rel,
                    }
                )
                break
    return pairs


def resolve_read_abs(root: Path, rel: str) -> Path | None:
    """Return first existing absolute file among candidates, or None."""
    for cand in candidate_read_paths(root, rel):
        if cand.is_file():
            return cand
    return None


def resolve_read(root: Path, rel: str) -> str | None:
    """Return vault-relative path of first existing candidate, or None."""
    abs_path = resolve_read_abs(root, rel)
    if abs_path is None:
        return None
    return vault_rel(root, abs_path)


def vault_rel(root: Path, abs_path: Path) -> str:
    return abs_path.resolve().relative_to(root.resolve()).as_posix()


def preferred_write_rel(rel: str, *, create: bool = False) -> str:
    """
    Canonical write path for a normalized knowledge rel.

    PARA paths stay as-is. Legacy/bare paths normalize to PARA
    (Inbox or Work/ResumePoints). ``create`` kept for API compatibility.
    """
    del create  # normalize already picks PARA destinations
    p = normalize_api_path(rel)
    if is_legacy_kb_path(p):
        suffix = p[len(LEGACY_KB_NOTES) + 1 :]
        if suffix.startswith("ResumePoints/"):
            return f"10-Work/{suffix}"
        return f"00-Inbox/{suffix}"
    return p


def resolve_write(root: Path, rel: str, *, create: bool = False) -> Path:
    """
    Absolute path to write.

    If the note already exists, write to the existing file.
    If creating, write under the preferred PARA path.
    """
    existing = resolve_read_abs(root, rel)
    if existing is not None and not create:
        return existing
    if existing is not None and create:
        return existing

    dest_rel = preferred_write_rel(rel, create=True)
    return abs_under_root(root, dest_rel)


def iter_knowledge_md(root: Path) -> Iterator[tuple[str, Path]]:
    """Yield (vault_rel, abs_path) for all PARA knowledge .md files."""
    seen_rels: set[str] = set()

    for area in PARA_AREAS:
        base = root / area
        if not base.is_dir():
            continue
        for path in sorted(base.rglob("*.md")):
            if path.name.startswith(".") or ".sync-conflict" in path.name:
                continue
            if is_chrome_basename(path.name):
                continue
            rel = vault_rel(root, path)
            if rel in seen_rels:
                continue
            seen_rels.add(rel)
            yield rel, path


def build_knowledge_tree(root: Path, *, section: str | None = None) -> dict[str, Any]:
    """Nested tree spanning PARA areas only.

    ``section`` optionally restricts to 'kb' (30-Knowledge) or 'notes'
    (all other PARA areas). None (default) returns everything.
    Chrome basenames (CLAUDE.md, .gitkeep, README.md) are omitted; MOCs kept.
    """
    section = validate_section(section)

    def empty_dir(path_label: str) -> dict[str, Any]:
        return {"path": path_label, "type": "dir", "children": []}

    def walk(directory: Path, vault_prefix: str) -> dict[str, Any]:
        children: list[dict[str, Any]] = []
        if not directory.is_dir():
            return empty_dir(vault_prefix)
        for child in sorted(
            directory.iterdir(), key=lambda p: (not p.is_dir(), p.name.lower())
        ):
            if child.name.startswith(".") or ".sync-conflict" in child.name:
                continue
            if is_chrome_basename(child.name):
                continue
            child_vault = f"{vault_prefix}/{child.name}" if vault_prefix else child.name
            if child.is_dir():
                node = walk(child, child_vault)
                if node["children"]:
                    children.append(node)
            elif child.suffix == ".md":
                children.append(
                    {"path": child_vault, "name": child.name, "type": "file"}
                )
        return {"path": vault_prefix, "type": "dir", "children": children}

    areas = PARA_AREAS
    if section == SECTION_KB:
        areas = (KB_AREA,)
    elif section == SECTION_NOTES:
        areas = NOTES_AREAS

    top: list[dict[str, Any]] = []
    for area in areas:
        area_path = root / area
        if area_path.is_dir():
            node = walk(area_path, area)
            if node["children"] or area_path.is_dir():
                top.append(node)

    return {
        "path": "knowledge",
        "type": "dir",
        "children": top,
    }


def list_knowledge_files(root: Path) -> list[str]:
    return [rel for rel, _ in iter_knowledge_md(root)]

File: chronicle-pc/server/src/path_map.rs
//! PARA knowledge path map (path_map.py): normalization, validation, tree.

use std::collections::BTreeSet;
use std::path::{Path, PathBuf};

use regex::Regex;

use crate::errors::ChronicleError;

pub const PARA_AREAS: [&str; 5] = ["00-Inbox", "10-Work", "20-Personal", "30-Knowledge", "90-Archive"];
pub const KB_AREA: &str = "30-Knowledge";
pub const NOTES_AREAS: [&str; 4] = ["00-Inbox", "10-Work", "20-Personal", "90-Archive"];
pub const SECTION_KB: &str = "kb";
pub const SECTION_NOTES: &str = "notes";
pub const LEGACY_KB_NOTES: &str = "kb/notes";
pub const CHROME_BASENAMES: [&str; 3] = ["CLAUDE.md", ".gitkeep", "README.md"];
pub const MACHINE_EXCLUDE_DIRS: [&str; 8] =
    ["index", "brain", "_capture", "_attachments", "entries", "img", "audio", ".stfolder"];

pub const MIGRATE_HINT: &str = "Legacy kb/notes/ dual-read is retired. Run `chronicle cutover-kb --apply --i-have-backup`, then use PARA paths (e.g. 30-Knowledge/…, 00-Inbox/…).";

const SAFE_REL: &str = r"^[A-Za-z0-9._\- /]+$";

pub fn is_chrome_basename(name: &str) -> bool {
    CHROME_BASENAMES.contains(&name)
}

pub fn _norm(path: &str) -> String {
    let cleaned = path.trim().trim_start_matches('/').replace('\\', "/");
    let mut collapsed = String::with_capacity(cleaned.len());
    let mut prev_slash = false;
    for ch in cleaned.chars() {
        if ch == '/' {
            if !prev_slash {
                collapsed.push(ch);
            }
            prev_slash = true;
        } else {
            collapsed.push(ch);
            prev_slash = false;
        }
    }
    collapsed
}

pub fn is_para_prefix(rel: &str) -> bool {
    PARA_AREAS.iter().any(|a| rel == *a || rel.starts_with(&format!("{a}/")))
}

pub fn is_legacy_kb_path(rel: &str) -> bool {
    rel == LEGACY_KB_NOTES || rel.starts_with("kb/notes/")
}

pub fn section_for(rel: &str) -> Option<&'static str> {
    if rel == KB_AREA || rel.starts_with(&format!("{KB_AREA}/")) {
        Some(SECTION_KB)
    } else if is_para_prefix(rel) {
        Some(SECTION_NOTES)
    } else {
        None
    }
}

pub fn default_create_area(section: &str) -> &'static str {
    if section == SECTION_KB { KB_AREA } else { "00-Inbox" }
}

pub fn validate_section(s: Option<&str>) -> Result<Option<String>, ChronicleError> {
    match s {
        None | Some("") => Ok(None),
        Some(v) if v == SECTION_KB || v == SECTION_NOTES => Ok(Some(v.to_string())),
        Some(other) => Err(ChronicleError::msg(format!(
            "section must be 'kb' or 'notes', got '{other}'"
        ))),
    }
}

/// normalize_api_path — passthrough rules + bare-path area assignment.
pub fn normalize_api_path(path: &str) -> Result<String, ChronicleError> {
    let p = _norm(path);
    if p.is_empty() {
        return Err(ChronicleError::msg("path must be a .md file"));
    }
    if is_legacy_kb_path(&p) {
        return Ok(p);
    }
    if is_para_prefix(&p) {
        return Ok(p);
    }
    if p == LEGACY_KB_NOTES {
        return Err(ChronicleError::msg("path must be a .md file"));
    }
    if p.starts_with("ResumePoints/") {
        return Ok(format!("10-Work/{p}"));
    }
    Ok(format!("00-Inbox/{p}"))
}

/// validate_knowledge_rel — full chain, exact error strings.
pub fn validate_knowledge_rel(rel: &str) -> Result<String, ChronicleError> {
    let p = normalize_api_path(rel)?;
    if is_legacy_kb_path(&p) {
        return Err(ChronicleError::msg(
            "legacy kb/notes/ path retired; run chronicle cutover-kb and use a PARA path",
        ));
    }
    if p.split('/').any(|c| c == "..") || p.contains('\0') || p.starts_with("../") || p.contains("/../") {
        return Err(ChronicleError::msg("invalid note path"));
    }
    let safe = Regex::new(SAFE_REL).unwrap();
    if !safe.is_match(&p) {
        return Err(ChronicleError::msg("note path has invalid characters"));
    }
    if !p.ends_with(".md") {
        return Err(ChronicleError::msg("note path must end with .md"));
    }
    if !is_para_prefix(&p) {
        return Err(ChronicleError::msg("path is not under a PARA knowledge area"));
    }
    Ok(p)
}

pub fn assert_path_allowed_for_section(rel: &str, section: Option<&str>) -> Result<(), ChronicleError> {
    let Some(section) = section else { return Ok(()) };
    let allowed = match section {
        SECTION_KB => rel == KB_AREA || rel.starts_with(&format!("{KB_AREA}/")),
        SECTION_NOTES => NOTES_AREAS.iter().any(|a| rel == *a || rel.starts_with(&format!("{a}/"))),
        _ => false,
    };
    if allowed {
        Ok(())
    } else {
        Err(ChronicleError::msg(format!(
            "path '{rel}' is outside section '{section}' (kb → 30-Knowledge/; notes → Inbox/Work/Personal/Archive)"
        )))
    }
}

/// preferred_write_rel: legacy kb/notes/SUFFIX → 10-Work (ResumePoints) or 00-Inbox.
pub fn preferred_write_rel(rel: &str) -> String {
    let p = match normalize_api_path(rel) {
        Ok(p) => p,
        Err(_) => return rel.to_string(),
    };
    if let Some(suffix) = p.strip_prefix("kb/notes/") {
        if suffix.starts_with("ResumePoints/") {
            return format!("10-Work/{suffix}");
        }
        return format!("00-Inbox/{suffix}");
    }
    p
}

fn abs_under_root(root: &Path, rel: &str) -> Result<PathBuf, ChronicleError> {
    let abs = root.join(rel);
    let root_resolved = root.canonicalize().unwrap_or_else(|_| root.to_path_buf());
    let resolved = abs.canonicalize().unwrap_or(abs.clone());
    if !resolved.starts_with(&root_resolved) && !abs.starts_with(&root_resolved) {
        return Err(ChronicleError::msg("note path escapes vault root"));
    }
    Ok(abs)
}

/// resolve_read_abs: existing file wins; legacy → None (caller emits 410).
pub fn resolve_read_abs(root: &Path, rel: &str) -> Result<Option<PathBuf>, ChronicleError> {
    let p = normalize_api_path(rel)?;
    if is_legacy_kb_path(&p) {
        return Ok(None);
    }
    let candidate = abs_under_root(root, &p)?;
    Ok(if candidate.is_file() { Some(candidate) } else { None })
}

/// resolve_write: existing file always wins; creates go to preferred rel.
pub fn resolve_write(root: &Path, rel: &str) -> Result<PathBuf, ChronicleError> {
    let p = normalize_api_path(rel)?;
    if let Some(existing) = resolve_read_abs(root, rel)? {
        return Ok(existing);
    }
    abs_under_root(root, &preferred_write_rel(&p))
}

fn skip_name(name: &str) -> bool {
    name.starts_with('.') || name.contains(".sync-conflict") || is_chrome_basename(name)
}

/// iter_knowledge_md: PARA areas in order, sorted rglob, dedup by rel.
pub fn iter_knowledge_md(root: &Path) -> Vec<(String, PathBuf)> {
    let mut seen = BTreeSet::new();
    let mut out = Vec::new();
    for area in PARA_AREAS {
        let dir = root.join(area);
        if !dir.is_dir() {
            continue;
        }
        let mut files: Vec<PathBuf> = Vec::new();
        collect_md_recursive(&dir, &mut files);
        files.sort();
        for f in files {
            if let Ok(rel) = f.strip_prefix(root) {
                let rel_str = rel.to_string_lossy().replace('\\', "/");
                let name = f.file_name().map(|n| n.to_string_lossy().to_string()).unwrap_or_default();
                if skip_name(&name) {
                    continue;
                }
                if seen.insert(rel_str.clone()) {
                    out.push((rel_str, f));
                }
            }
        }
    }
    out
}

fn collect_md_recursive(dir: &Path, out: &mut Vec<PathBuf>) {
    crate::paths::walk_files_filtered(dir, out, 0, &|_p, name| {
        !skip_name(name) && name.ends_with(".md")
    });
}

#[derive(Debug, Clone, serde::Serialize)]
pub struct TreeNode {
    pub path: String,
    #[serde(rename = "type")]
    pub kind: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub children: Option<Vec<TreeNode>>,
}

/// build_knowledge_tree — dirs-first sort, childless-dir pruning, chrome hidden.
pub fn build_knowledge_tree(root: &Path, section: Option<&str>) -> TreeNode {
    let mut children = Vec::new();
    for area in PARA_AREAS {
        let dir = root.join(area);
        if !dir.is_dir() {
            continue;
        }
        if let Some(node) = build_dir_node(root, &dir, section) {
            children.push(node);
        }
    }
    TreeNode {
        path: "knowledge".into(),
        kind: "dir".into(),
        name: None,
        children: Some(children),
    }
}

fn build_dir_node(root: &Path, dir: &Path, section: Option<&str>) -> Option<TreeNode> {
    build_dir_node_depth(root, dir, section, 0)
}

fn build_dir_node_depth(root: &Path, dir: &Path, section: Option<&str>, depth: usize) -> Option<TreeNode> {
    if depth >= crate::paths::MAX_WALK_DEPTH {
        return None;
    }
    let mut dirs: Vec<TreeNode> = Vec::new();
    let mut files: Vec<TreeNode> = Vec::new();
    for (path, is_dir) in crate::paths::list_children(dir) {
        let name = path.file_name().map(|n| n.to_string_lossy().to_string()).unwrap_or_default();
        if skip_name(&name) {
            continue;
        }
        let rel = path
            .strip_prefix(root)
            .map(|r| r.to_string_lossy().replace('\\', "/"))
            .unwrap_or_default();
        if is_dir {
            if let Some(child) = build_dir_node_depth(root, &path, section, depth + 1) {
                dirs.push(child);
            }
        } else if name.ends_with(".md") {
            if let Some(sec) = section_for(&rel) {
                if section.map(|s| s != sec).unwrap_or(false) {
                    continue;
                }
                files.push(TreeNode { path: rel, kind: "file".into(), name: Some(name), children: None });
            }
        }
    }
    dirs.sort_by(|a, b| a.path.to_lowercase().cmp(&b.path.to_lowercase()));
    files.sort_by(|a, b| a.name.as_deref().unwrap_or("").to_lowercase().cmp(&b.name.as_deref().unwrap_or("").to_lowercase()));
    let mut kids = dirs;
    kids.extend(files);
    if kids.is_empty() {
        // Existing-but-childless areas still appear; deeper empty dirs are pruned.
        let rel = dir.strip_prefix(root).map(|r| r.to_string_lossy().replace('\\', "/")).unwrap_or_default();
        if PARA_AREAS.contains(&rel.as_str()) {
            return Some(TreeNode { path: rel, kind: "dir".into(), name: None, children: Some(vec![]) });
        }
        return None;
    }
    let rel = dir.strip_prefix(root).map(|r| r.to_string_lossy().replace('\\', "/")).unwrap_or_default();
    Some(TreeNode { path: rel, kind: "dir".into(), name: None, children: Some(kids) })
}

#[cfg(all(test, unix))]
mod tests {
    use super::*;

    #[test]
    fn iter_knowledge_md_skips_symlinked_dirs() {
        let tmp = tempfile::tempdir().unwrap();
        let root = tmp.path();
        let inbox = root.join("00-Inbox");
        std::fs::create_dir_all(&inbox).unwrap();
        std::fs::write(inbox.join("note.md"), "# note").unwrap();
        // Self-cycle and ancestor-cycle: walkers must not descend into either,
        // and must not emit duplicate rel paths for files reachable through them.
        std::os::unix::fs::symlink(&inbox, inbox.join("self")).unwrap();
        std::os::unix::fs::symlink(root, inbox.join("rootlink")).unwrap();

        let files = iter_knowledge_md(root);
        let rels: Vec<&str> = files.iter().map(|(r, _)| r.as_str()).collect();
        assert_eq!(rels, vec!["00-Inbox/note.md"], "unexpected: {rels:?}");
    }
}
